File: android/src/main/kotlin/com/example/flutter_rtmp_streamer/DartMessenger.kt
/*
 * Copyright (C) 2022 kuzalex.
 *
 */

package com.example.flutter_rtmp_streamer


import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink


//
// This EventChannel is used to stream data from plugin to flutter.
//

class DartMessenger(messenger: BinaryMessenger, name: String) {
    private var eventSink: EventSink? = null

    public val handler: Handler by lazy {
        Handler(Looper.getMainLooper())
    }

    fun send(eventType: String, args: Map<String, String>) {
        handler.post {_send(eventType, args)}
    }

    private fun _send(eventType: String, args: Map<String, String>) {
        if (eventSink == null) {
            return
        }
        var data: MutableMap<String, String> = args.toMutableMap()
        data["eventType"] = eventType

        eventSink!!.success(data)
    }

    init {
        assert(messenger != null);
        EventChannel(messenger, name)
                .setStreamHandler(
                        object : EventChannel.StreamHandler {
                            override fun onListen(arguments: Any?, sink: EventSink) {
                                eventSink = sink
                            }

                            override fun onCancel(arguments: Any?) {
                                eventSink = null
                            }
                        })
    }
}

File: ios/Classes/DartMessenger.swift
//
//  DartMessenger.swift
//  flutter_rtmp_streamer
//
//  Created by kuzalex on 5/10/22.
//

import Foundation
import Flutter
import UIKit

class DartMessenger : NSObject, FlutterStreamHandler {
    private var _eventSink: FlutterEventSink?

    init(messenger: FlutterBinaryMessenger, name: String) {
        super.init()

        let eventCnannel = FlutterEventChannel.init(name: name, binaryMessenger: messenger)


        eventCnannel.setStreamHandler(self)
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        _eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        _eventSink = nil
        return nil
    }

    func send(eventType: String, args: [String:Any]) {
        guard let eventSink = _eventSink else {
            return
        }
        var data:[String:Any] = args
        data["eventType"] = eventType

        eventSink(data)

    }
}
